netsim/modules/lag.py

```diff
@@ -50,186 +63,207 @@ def check_lag_config(node: str, linkname: str, topology: Box) -> bool:
   return True
 
 """
-check_same_pair - Verifies that the given member connects the same pair of nodes as the first
+check_mlag_support - check if the given node supports mlag
 """
-def check_same_pair(first_pair: list[str], member: Box) -> bool:
-  others = { n.node for n in member.interfaces if n.node not in first_pair }
-  if others:
-    log.error(f'All member links must connect the same pair of nodes({first_pair}), found {others}',
+def check_mlag_support(node: str, linkname: str, topology: Box) -> bool:
+  _n = topology.nodes[node]
+  features = devices.get_device_features(_n,topology.defaults)
+  if not features.lag.get('mlag',False):
+    log.error(f'Node {_n.name} ({_n.device}) does not support MLAG, cannot be part of peerlink or M-side of LAG {linkname}',
       category=log.IncorrectValue,
       module='lag')
     return False
   return True
 
+"""
+normalized_members - builds a normalized list of lag member links, checking various conditions
+"""
+def normalized_members(l: Box, topology: Box, peerlink: bool = False) -> list:
+  members = []                                      # Build normalized list of members
+  _name = "peerlink" if peerlink else "lag"
+  for idx,member in enumerate(l.lag.members):
+    member = links.adjust_link_object(member,f'{l._linkname}.{_name}[{idx+1}]',topology.nodes)
+    if 'lag' in member:                             # Catch potential sources for inconsistency
+      if peerlink or ('ifindex' not in member.lag): # ...but allow for custom bond numbering
+        log.error(f'LAG attributes must be configured on the link, not member interface {member._linkname}: {member.lag}',
+                  category=log.IncorrectAttr,
+                  module='lag')
+        return []
+    if len(member.interfaces)!=2:                   # Check that there are exactly 2 nodes involved
+      log.error(f'Link {member._linkname} in LAG {l.lag.ifindex} must have exactly 2 nodes',
+                category=log.IncorrectValue,
+                module='lag')
+      return []
+    if member.interfaces[0].node==member.interfaces[1].node:
+      log.error(f'Link {member._linkname} in LAG {l.lag.ifindex} must have exactly 2 different nodes',
+                category=log.IncorrectValue,
+                module='lag')
+      return []
+    for i in member.interfaces:                     # Check that they all support LAG
+      if not check_lag_config(i.node,member._linkname,topology):
+        return []
+    members.append(member)
+  return members
+
 """
 create_lag_member_links -- expand lag.members for link l and create physical p2p links
 """
 def create_lag_member_links(l: Box, topology: Box) -> None:
```

Comment:
Is this a function with four (or so) embedded functions? Why did you use embedded functions? It would be much easier to read if you would take those functions out of this one.

Reply:
The embedded functions are only used from this context, and use many of the same arguments. I have moved them out (except ```analyze_lag```)